File: fix_nb.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Fix the split notebook JSON file."""

if __name__ == '__main__':
    import json
    import sys
    import os
    
    file_path = r'C:\laragon\www\anemo-ai-v3\scripts\ml\AnemoAI_Training_Colab.ipynb'
    
    try:
        # Read the file
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        # Find the first valid complete JSON object by counting braces
        depth = 0
        end_pos = 0
        for i, ch in enumerate(content):
            if ch == '{':
                depth += 1
            elif ch == '}':
                depth -= 1
                if depth == 0:
                    end_pos = i + 1
                    break
        
        valid_content = content[:end_pos]
        
        # Verify it's valid JSON
        try:
            json.loads(valid_content)
            print("✓ Valid JSON verified")
        except json.JSONDecodeError as e:
            print(f"✗ Invalid JSON: {e}")
            sys.exit(1)
        
        # Write back
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(valid_content)
        
        print(f'✓ Success: wrote {len(valid_content)} chars, original was {len(content)} chars')
        print(f'✓ Removed {len(content) - len(valid_content)} chars of garbage')
        print(f'✓ File: {file_path}')
        
    except Exception as e:
        print(f'✗ Error: {e}')
        sys.exit(1)

File: fix_nb.js
const fs = require('fs');
const path = require('path');

const filePath = path.join(__dirname, 'scripts', 'ml', 'AnemoAI_Training_Colab.ipynb');

console.log(`Reading: ${filePath}`);

// Read the file
const content = fs.readFileSync(filePath, 'utf-8');

// Find the first valid complete JSON object by counting braces
let depth = 0;
let endPos = 0;
for (let i = 0; i < content.length; i++) {
    const ch = content[i];
    if (ch === '{') {
        depth++;
    } else if (ch === '}') {
        depth--;
        if (depth === 0) {
            endPos = i + 1;
            break;
        }
    }
}

const validContent = content.substring(0, endPos);

// Verify it's valid JSON
try {
    JSON.parse(validContent);
    console.log('✓ Valid JSON verified');
} catch (e) {
    console.error('✗ Invalid JSON:', e.message);
    process.exit(1);
}

// Write back
fs.writeFileSync(filePath, validContent, 'utf-8');

console.log(`✓ Success: wrote ${validContent.length} chars, original was ${content.length} chars`);
console.log(`✓ Removed ${content.length - validContent.length} chars of garbage`);
console.log(`✓ File: ${filePath}`);
